openssl/src/x509/store.rs

```diff
@@ -94,6 +94,80 @@ impl X509StoreBuilderRef {
     pub fn set_default_paths(&mut self) -> Result<(), ErrorStack> {
         unsafe { cvt(ffi::X509_STORE_set_default_paths(self.as_ptr())).map(|_| ()) }
     }
+
+    /// Adds a lookup method to the store.
+    ///
+    /// This corresponds to [`X509_STORE_add_lookup`].
+    ///
+    /// [`X509_STORE_add_lookup`]: https://www.openssl.org/docs/man1.1.1/man3/X509_STORE_add_lookup.html
+    pub fn add_lookup(
+        &mut self,
+        method: &X509LookupMethodRef,
+    ) -> Result<&mut X509LookupRef, ErrorStack> {
+        let lookup = unsafe { ffi::X509_STORE_add_lookup(self.as_ptr(), method.as_ptr()) };
+        cvt_p(lookup).map(|ptr| unsafe { X509LookupRef::from_ptr_mut(ptr) })
+    }
+}
+
+foreign_type_and_impl_send_sync! {
+    type CType = ffi::X509_LOOKUP;
+    fn drop = ffi::X509_LOOKUP_free;
+
+    /// Information used by an `X509Store` to look up certificates and CRLs.
+    pub struct X509Lookup;
+    /// Reference to an `X509Lookup`.
+    pub struct X509LookupRef;
+}
+
+impl X509Lookup {
+    /// Lookup method that loads certificates and CRLs on demand and caches
+    /// them in memory once they are loaded. As of OpenSSL 1.0.0, it also
+    /// checks for newer CRLs upon each lookup, so that newer CRLs are used as
+    /// soon as they appear in the directory.
+    ///
+    /// This corresponds to [`X509_LOOKUP_hash_dir`].
+    ///
+    /// [`X509_LOOKUP_hash_dir`]: https://www.openssl.org/docs/man1.1.0/crypto/X509_LOOKUP_hash_dir.html
+    pub fn hash_dir() -> &'static X509LookupMethodRef {
+        unsafe { X509LookupMethodRef::from_ptr(ffi::X509_LOOKUP_hash_dir()) }
+    }
+}
+
+impl X509LookupRef {
+    /// Specifies a directory from which certificates and CRLs will be loaded
+    /// on-demand. Must be used with `X509Lookup::hash_dir`.
```

Comment:
This is a potential soundness concern if any other lookup method uses CTL codes that collide with X509_L_ADD_DIR

Reply:
If I'm understanding the concern correctly, would it make sense to make `X509Lookup` generic over a type indicating which CTL codes it supports? For instance, `hash_dir` would return something like `X509LookupRef<Directory>`, and `add_dir` would only be implemented on `X509LookupRef<Directory>`

Reply:
Yeah, that would solve it (though I'd call the type `HashDir` rather than `Directory` to match the constructor).

Reply:
Alright, implemented this idea. Decided to name it `AddDir` after `X509_L_ADD_DIR` since I think it makes more sense to have the generic parameter correspond to the control commands supported rather than the exact method it came from. This might be problematic if any lookup methods support multiple control commands--in that case, maybe we'd need traits indicating support for each--but I think it should be sufficient for now.

Reply:
> This is a potential soundness concern if any other lookup method uses CTL codes that collide with X509_L_ADD_DIR

Rereading this, I might've taken a step in the wrong direction here. Is what you're saying that two lookup methods might interpret the same command to mean different things? In that case, I'll switch the marker type back to `HashDir`, since that's what we're actually looking for.

Reply:
Yeah, we'll want the parameter to indicate what kind of lookup algorithm it is (HashDir) in this case, since that's what determines how the X509_LOOKUP_ctrl behaves. There could be other ctrl APIs that the hash_dir method supports and they'd all end up under this type parameter for example.

Reply:
Gotcha, done